### DISCUSSION TITLE
Magento 2 recipe maintenance flag in shared files

### DISCUSSION THREAD

**Discussion by lfolco** [deployphp/deployer#2940] [Help needed] [answered]
Currently the Magento 2 recipe has the `var/.maintenance.ip` file under `shared_files`; should it also include `var/.maintenance.flag` in case the user does not specify allowed IP addresses?

**Comment by peterjaap** [comment]
Yes, good point.

**Comment by peterjaap** [comment, marked as answer]
Done https://github.com/deployphp/deployer/commit/abfb9bb235ec57d85d2365a936bc44f884db8514

**Reply by lfolco** [reply]
Thank you!

**Comment by valguss** [comment]
Sorry to reopen this but I think this change means that any deployment will result in the site being set to maintenance mode as the `.maintenance.flag` file will be created if it currently doesn't exist

**Reply by sprankhub** [reply]
Exactly. Any deployment now leads to the shop being in maintenance mode. What is the reasoning behind this, @peterjaap? Can we remove `var/.maintenance.flag` from the shared files again, please?

**Reply by lfolco** [reply]
I remember asking for this; when you wanted maintenance mode it didn't work without this (my hosting provider doesn't handle zero-downtime deployments). Anyway of making a conditional check?

**Reply by peterjaap** [reply]
@valguss you're right, I hadn't noticed that implication - I didn't know Deployer creates a file when it doesn't exist - I *assumed* (my bad) it only creates a symlink when the file actually exists.

@lfolco I don't see how a hosting provider could have influence on zero-downtime deployments not working? Could you elaborate? Maybe we can solve that issue for you so you don't need this setting. For now I'm reverting this change.

**Reply by lfolco** [reply]
They don't support it with their load balancers (MageMojo). Until I turned on maintenance mode, most of my deploys broke with customers getting the wrong files. Is there a way to set it conditionally?

**Reply by lfolco** [reply]
Essentially they have a 2-3 minute process where files are copied. If the site is up during that time things break.

**Reply by valguss** [reply]
That sounds like you have it set up incorrectly or caches are not getting flushed correctly afterwards. rc5 introduced a flag for zero downtime though: https://github.com/deployphp/deployer/releases/tag/v7.0.0-rc.5

**Reply by valguss** [reply]
ROFL, by you lol

**Reply by lfolco** [reply]
Believe me, I've been back and forth with them over this, the only solution that doesn't result in weird errors after every deploy is putting it in maintenance mode. The issue is, as it works now (without the flag in shared), maintenance mode won't work at all, so reverting it makes that update I made useless.

**Reply by valguss** [reply]
I guess really you need to explain the functionality you require @lfolco as I'm not quite seeing it currently

**Reply by peterjaap** [reply]
@lfolco you could just add it to your own config. It doesn't have to be in the core recipe. In your `deploy.php`, just add;

```
add('shared_files, ['var/.maintenance.flag']);
```

**Reply by valguss** [reply]
yeah, almost sounds like you just need to do:
```
before('deploy', 'magento:maintenance:enable');
```
or similar so you force into MM at the start of a deployment

**Reply by lfolco** [reply]
I'll look into a PR that does that conditionally, because otherwise the maintenance mode changes I made will no longer work.

**Reply by valguss** [reply]
aah, so basically if MM is enabled already, then you need it to persist through the deployment. Maybe do a test for .maintenance.flag in the current release:
```
test('[ -f {{current_path}}/var/.maintenance.flag ]');
```
think that'll path to the currently live version

**Reply by lfolco** [reply]
That's an idea, I'll take it for a spin. Thanks!

**Comment by peterjaap** [comment]
Reverted this change in https://github.com/deployphp/deployer/commit/297e57dc3c92938c8440f6ee616bfef9bc37b7ed

**Reply by valguss** [reply]
thanks @peterjaap

**Comment by lfolco** [comment]
Please find a different way to handle this, since without that flag in shared, you cannot set maintenance mode at all when deploying.

**Reply by peterjaap** [reply]
Sure it does. It sets the current release in maintenance mode, then runs setup:upgrade, then switches the symlink and lastly removes the maintenance flag.

**Reply by lfolco** [reply]
I was never able to get it to work; my site was free for all to browse when it was supposed to be in maintenance mode until we added the flag to shared.

**Reply by lfolco** [reply]
Maybe it's MageMojo's setup, I don't know. Their Stratus has been a problem since we were forced to switch to it.

**Reply by peterjaap** [reply]
That definitely sounds like it's in their setup somewhere. This has worked for us across multiple hosting providers as well as our own review environments on Google Cloud.

**Reply by valguss** [reply]
We've had Similar with Sonassi but that was because we needed to clear the opcache, so I created a separate task that sorts that out. Maybe the same here and the [CacheTool recipe](https://deployer.org/docs/7.x/contrib/cachetool) doesn't do it for you?

**Reply by valguss** [reply]
Ahh, that recipe should work in your case: https://magemojo.com/kb/best-practices/opcache/
